airflow/models/xcom_arg.py

```diff
@@ -0,0 +1,155 @@
+        self._operator = operator
+
+        # _key is used for xcompull (key in the XCom table)
+        # _runtime_sub_keys are used for runtime access of deserialized value from XCom data provider
+        self._key = key
+        self._runtime_sub_keys = runtime_sub_keys or []
+
+    def __eq__(self, other):
+        return (self._operator == other.operator
+                and self.key == other.key
+                and self.runtime_sub_keys == other.runtime_sub_keys)
+
+    def __lshift__(self, other):
+        """
+        Implements xcomresult << op
+        """
+        self.set_upstream(other)
+        return self
+
+    def __rshift__(self, other):
+        """
+        Implements xcomresult << op
+        """
+        self.set_downstream(other)
+        return self
+
+    @property
+    def operator(self) -> BaseOperator:
+        """Operator"""
+        return self._operator
+
+    @property
+    def key(self):
+        return self._key
+
+    @property
+    def runtime_sub_keys(self):
+        return self._runtime_sub_keys
+
+    def set_upstream(self, task_or_task_list: Union[BaseOperator, List[BaseOperator]]):
+        """
+        Proxy to underlying operator set_upstream method
+        """
+        self._operator.set_upstream(task_or_task_list)
+
+    def set_downstream(
+        self, task_or_task_list: Union[BaseOperator, List[BaseOperator]]
+    ):
+        """
+        Proxy to underlying operator set_downstream method
+        """
+        self._operator.set_downstream(task_or_task_list)
+
+    def resolve(self, context: Dict) -> Any:
+        """
+        Pull XCom value for the existing arg.
+        this function likely to run during at op.execute() context
+        """
+        resolved_value = self._operator.xcom_pull(
+            context=context,
+            task_ids=[self._operator.task_id],
+            key=self.key,
+            dag_id=self._operator.dag.dag_id,
+        )
+        if not resolved_value:
+            raise AirflowException(
+                f'XComArg result from {self._operator.task_id} at {self._operator.dag.dag_id} '
+                f'with key="{self.key}"" is not found!')
+        resolved_value = resolved_value[0]
+
+        # now we can follow on user defined dereference path
+        if len(self._runtime_sub_keys) > 1:
+            for key in self._runtime_sub_keys:
+                resolved_value = resolved_value[key]
+        return resolved_value
+
+    def __getitem__(self, key: Any) -> 'XComArg':
+        """Return an XComArg that will access the following """
+        return XComArg(operator=self._operator,
+                       key=self.key,
+                       runtime_sub_keys=self._runtime_sub_keys + [key])
+
+    def __str__(self):
+        """
+        Backward compatibility for old-style jinja used in Airflow Operators
+        **Example**: to use XArg at BashOperator::
+            BashOperator(cmd=f"... { xcomarg } ...")
+        :return:
```

Comment:
Is it real backward compatibility? I mean, previously we had to use `{{ task_instance.xcom_pull.. }}` and it still will work right? So, isn't using `xcomarg` as an input a new feature?

Reply:
xcomarg itself is the new feature, and you are right, that xcomarg can be access via {{task_instance..}}. but that makes the user aware of xargcom internal implementation,  till now all `xcom_push` was done mostly explicitly by user code, now there is some code that pushes, but pull implementation is on the user.  using same logic we don't need to support << and >>

I will remove __str__ implementation for now.

Reply:
Let's keep it! I like it, my main concern is about "Backward compatibility" part in docstring. Once we remove it I think it's a nice feature